### PR TITLE
Query: Client Eval DateTime arithmetic operation for Sqlite/SqlServer

### LINKED ISSUE #10656: Incorrect translation of date substraction.
```
var delays = await db
    .Set<Job>()
    .Select(j => (j.StartJob - j.CreateDate).TotalHours)
    .ToListAsync();
```

Gets translated into:
```
SELECT [j].[StartJob] - [j].[CreateDate]
FROM [Job] AS [j]
WHERE [j].[IsDeleted] = 0
```

Which in turn throws:
```
System.Data.SqlClient.SqlException (0x80131904): Operand data type datetime2 is invalid for subtract operator.
   at System.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__108_0(Task`1 result)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.<ExecuteAsync>d__17.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable`1.AsyncEnumerator.<BufferlessMoveNext>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.EntityFrameworkCore.Storage.Internal.SqlServerExecutionStrategy.<ExecuteAsync>d__7`2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable`1.AsyncEnumerator.<MoveNext>d__9.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.<MoveNext>d__5.MoveNext()
ClientConnectionId:c7a3cc04-ddbf-4aab-9957-7477083c909f
Error Number:8117,State:1,Class:16
```

This query should either be translated into DateDiff if possible or perform subtraction on client but definitely not throw.

## Further technical details
EF Core version: 2.0.1
Database Provider: Microsoft.EntityFrameworkCore.SQLServer
Operating system: Windows 10 1703
IDE: Visual Studio 2017

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #10656

### CODE REVIEW

File: src/EFCore.SqlServer/Query/ExpressionVisitors/Internal/SqlServerSqlTranslatingExpressionVisitor.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                case ExpressionType.Modulo:
+                    return IsDateTimeBasedOperation(visitedExpression)
+                        ? null
+                        : visitedExpression;
+            }
+
+            return visitedExpression;
+        }
+
+        private bool IsDateTimeBasedOperation(Expression expression)
```
Comment (anpete) [inline src/EFCore.SqlServer/Query/ExpressionVisitors/Internal/SqlServerSqlTranslatingExpressionVisitor.cs]: static?

File: src/EFCore.SqlServer/Query/ExpressionVisitors/Internal/SqlServerSqlTranslatingExpressionVisitor.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                    && _dateTimeDataTypes.Contains(typeMapping.StoreType))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private RelationalTypeMapping InferTypeMappingFromColumn(Expression expression)
```
Comment (anpete) [inline src/EFCore.SqlServer/Query/ExpressionVisitors/Internal/SqlServerSqlTranslatingExpressionVisitor.cs]: static?

File: src/EFCore.Sqlite.Core/Query/ExpressionVisitors/Internal/SqliteSqlTranslatingExpressionVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                        || typeMapping is SqliteDateTimeTypeMapping))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private RelationalTypeMapping InferTypeMappingFromColumn(Expression expression)
```
Comment (anpete) [inline src/EFCore.Sqlite.Core/Query/ExpressionVisitors/Internal/SqliteSqlTranslatingExpressionVisitor.cs]: Add method to base type?

File: src/EFCore.Sqlite.Core/Query/ExpressionVisitors/Internal/SqliteSqlTranslatingExpressionVisitor.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                case ExpressionType.Modulo:
+                    return IsDateTimeBasedOperation(visitedExpression)
+                        ? null
+                        : visitedExpression;
+            }
+
+            return visitedExpression;
+        }
+
+        private bool IsDateTimeBasedOperation(Expression expression)
```
Comment (anpete) [inline src/EFCore.Sqlite.Core/Query/ExpressionVisitors/Internal/SqliteSqlTranslatingExpressionVisitor.cs]: Make template method on base type?

File: src/EFCore.Sqlite.Core/Query/ExpressionVisitors/Internal/SqliteSqlTranslatingExpressionVisitor.cs (hunk 3 above)
Comment (smitpatel) [reply src/EFCore.Sqlite.Core/Query/ExpressionVisitors/Internal/SqliteSqlTranslatingExpressionVisitor.cs]: Would keep this local & private for now till we understand all interactions with value conversions.

**Comment by smitpatel** [comment]
Updated.

**Pull request merged**